autopush_rs/src/util/ddb_helpers.rs

```diff
@@ -102,3 +129,479 @@ macro_rules! ddb_item {
         }
     }
 }
+
+/// Direct representation of a DynamoDB Notification as we store it in the database
+/// Most attributes are optional
+#[derive(Default, Deserialize, PartialEq, Debug, Clone)]
+pub struct NotificationHeaders {
+    crypto_key: Option<String>,
+    encryption: Option<String>,
+    encryption_key: Option<String>,
+    encoding: Option<String>,
+}
+
+fn insert_to_map(map: &mut HashMap<String, String>, val: Option<String>, name: &str) {
+    if let Some(val) = val {
+        map.insert(name.to_string(), val);
+    }
+}
+
+impl From<NotificationHeaders> for HashMap<String, String> {
+    fn from(val: NotificationHeaders) -> HashMap<String, String> {
+        let mut map = HashMap::new();
+        insert_to_map(&mut map, val.crypto_key, "crypto_key");
+        insert_to_map(&mut map, val.encryption, "encryption");
+        insert_to_map(&mut map, val.encryption_key, "encryption_key");
+        insert_to_map(&mut map, val.encoding, "encoding");
+        map
+    }
+}
+
+#[derive(Default, Deserialize, PartialEq, Debug, Clone)]
+pub struct DynamoDbNotification {
+    // DynamoDB <Hash key>
+    uaid: Uuid,
+    // DynamoDB <Range key>
+    // Format:
+    //    Topic Messages:
+    //        01:{channel id}:{topic}
+    //    New Messages:
+    //        02:{timestamp int in microseconds}:{channel id}
+    chidmessageid: String,
+    // Magic entry stored in the first Message record that indicates the highest
+    // non-topic timestamp we've read into
+    current_timestamp: Option<u64>,
+    // Magic entry stored in the first Message record that indicates the valid
+    // channel id's
+    chids: Option<HashSet<String>>,
+    // Time in seconds from epoch
+    timestamp: Option<u32>,
+    // DynamoDB expiration timestamp per
+    //    https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/TTL.html
+    expiry: u32,
+    // TTL value provided by application server for the message
+    ttl: Option<u32>,
+    data: Option<String>,
+    headers: Option<NotificationHeaders>,
+    // This is the acknowledgement-id used for clients to ack that they have received the
+    // message. Some Python code refers to this as a message_id. Endpoints generate this
+    // value before sending it to storage or a connection node.
+    updateid: Option<String>,
+}
+
+struct RangeKey {
+    channel_id: Uuid,
+    topic: Option<String>,
+    sortkey_timestamp: Option<u64>,
+    legacy_version: Option<String>,
+}
+
+fn parse_sort_key(key: &str) -> Result<RangeKey> {
+    lazy_static! {
+        static ref RE: RegexSet = RegexSet::new(&[
+            r"^01:\S+:\S+$",
+            r"^02:\d+:\S+$",
+            r"^\S{3,}:\S+$",
+        ]).unwrap();
+    }
+    if !RE.is_match(key) {
+        return Err("Invalid chidmessageid".into()).into();
+    }
+    if key.starts_with("01:") {
```

Comment:
part of me wonders if this should be a match...

Reply:
I don't think strings can be destructured to partial match right now.